### PR TITLE
fix: grpc-web trailers

### LINKED ISSUE #10573: bug: grpc-web plugin causing missing trailers issue
### Current Behavior

When i use the grpc-web plugin and cors plugin to communicate from my fronted to my backend, the response I receive an error missing trailers. I've tried the same thing using envoy and it works. There seems to be some misconfiguration is the response body which does not allow the response to

### Expected Behavior
Screenshot 2023-11-28 at 10 52 01 AM

Envoy Response: (CORRECT AND DOES NOT CAUSE ISSUE)

Hello! Mohsin Javed� grpc-status:0
grpc-message:OK

Headers:

HTTP/1.1 200 OK
content-type: application/grpc-web+proto
grpc-accept-encoding: identity,deflate,gzip
grpc-encoding: identity
date: Fri, 24 Nov 2023 06:04:45 GMT
x-envoy-upstream-service-time: 12
access-control-allow-origin: http://localhost:58653/
access-control-expose-headers: custom-header-1,grpc-status,grpc-message
server: envoy
transfer-encoding: chunked

API SIX Response:: (INCORRECT 

AND CAUSES ISSUE MISSING TRAILERS)

Hello! World

Headers:
HTTP/1.1 200 OK
Date: Fri, 24 Nov 2023 06:13:48 GMT
Content-Type: application/grpc-web+proto
Transfer-Encoding: chunked
Connection: keep-alive
grpc-accept-encoding: identity,deflate,gzip
grpc-encoding: identity
Server: APISIX/3.6.0
Access-Control-Allow-Origin: *
Access-Control-Allow-Methods: *
Access-Control-Max-Age: 5
Access-Control-Expose-Headers: *
Access-Control-Allow-Headers: *

### Error Logs

Screenshot 2023-11-30 at 7 05 22 PM

### Steps to Reproduce

enable grpc-web and cors plugin on a route that communicates from any frontend / backend service that implemented a grpc service using a proto definition.

### Environment

Please take a clone of this https://github.com/m-jayy/grpc_with_envoy_and_apisix/pull/1 All steps to reproduce are mentioned in the README.md

This is a small local setup with node as the back end and flutter as the frontend

### PR DESCRIPTION
**Pull request by sheharyaar**
### Description

Fixes #10573  and closes #10613

### Checklist

- [X] I have explained the need for this PR and the problem it solves
- [X] I have explained the changes or the new features added to this PR
- [ ] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [ ] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

**Comment by sheharyaar** [comment]
This PR is not yet complete, need to add some features.
@Affanmir, please checkout this PR and confirm if this fixes your issue ?

**Comment by Vacant2333** [comment]
@sheharyaar hello~ have u test on your enviorment? u can add some testcase that can help us to review~~

**Comment by monkeyDluffy6017** [comment]
@sheharyaar Is this pr ready?

**Comment by sheharyaar** [comment]
Yes, the PR is ready @monkeyDluffy6017

**Comment by Affanmir** [comment]
> This PR is not yet complete, need to add some features. @Affanmir, please checkout this PR and confirm if this fixes your issue ?

Missing trailers issue fixed but path based testing cannot be done until the cors issue is resolved.

Access to XMLHttpRequest at 'https://XXXXXXXX/RANDOM.RANDOM.SERVICENAME.service.ServiceExternal/SERVICEMETHOF' from origin 'http://localhost:8080/' has been blocked by CORS policy: Request header field REDACTED_FIELD is not allowed by Access-Control-Allow-Headers in preflight response.
SUBDOMAON.DOMAIN/RANDOM.RANDOM.SERVICENAME.service.ServiceExternal/SERVICEMETHOD:1


The issue persistents even when using cors plugin alongside this plugin. Please replicate on your end

**Review by Affanmir — APPROVED** [review]

File: apisix/plugins/grpc-web.lua [hunk 1]
```diff
@@ -147,6 +151,49 @@ function _M.body_filter(conf, ctx)
...
+        trailer_buf = trailer_buf .. string.char(
+            bit.band(bit.rshift(len, 24), 0xff),
+            bit.band(bit.rshift(len, 16), 0xff),
+            bit.band(bit.rshift(len, 8), 0xff),
+            bit.band(len, 0xff)
+        )
+        -- n bytes: trailer
+        trailer_buf = trailer_buf .. grpc_web_trailer
+
+        if ctx.grpc_web_encoding == CONTENT_ENCODING_BINARY then
```
Comment (shreemaan-abhishek) [inline apisix/plugins/grpc-web.lua:187]: checking the encoding twice, can this be removed now? https://github.com/apache/apisix/blob/196461e2ec728a763ca13cbef337d9ce0fb2abc1/apisix/plugins/grpc-web.lua#L148C1-L153C8

Comment (sheharyaar) [reply apisix/plugins/grpc-web.lua:187]: This encoding check is for trailer only, the previous check is to base64encode the body (needed only if encoding is base64). This check is required as trailers is needed for both the cases and requires different handling strategy.

This executes only if the status is not pty or nil. The previous check executes every time for the body content.

File: apisix/plugins/grpc-web.lua [hunk 2]
```diff
@@ -147,6 +151,49 @@ function _M.body_filter(conf, ctx)
...
+        local len = #grpc_web_trailer
+
+        -- 1 byte: 0x80
+        local trailer_buf = string.char(0x80)
+        -- 4 bytes: length of the trailer
+        trailer_buf = trailer_buf .. string.char(
+            bit.band(bit.rshift(len, 24), 0xff),
+            bit.band(bit.rshift(len, 16), 0xff),
+            bit.band(bit.rshift(len, 8), 0xff),
+            bit.band(len, 0xff)
```
[the remark this answers is not in the document]
Comment (sheharyaar) [reply apisix/plugins/grpc-web.lua:182]: Yes, the length of the trailers part needs to be exact 4 bytes.

File: apisix/plugins/grpc-web.lua [hunk 3]
```diff
@@ -147,6 +151,49 @@ function _M.body_filter(conf, ctx)
...
+
+    Format for grpc-web trailer
+        1 byte: 0x80
+        4 bytes: length of the trailer
+        n bytes: trailer
+
+    --]]
+    local status = ctx.var.upstream_trailer_grpc_status
+    if status ~= "" and status ~= nil then
+        local status_str = "grpc-status:" .. tonumber(status)
```
[the remark this answers is not in the document]
Comment (sheharyaar) [reply apisix/plugins/grpc-web.lua:170]: I don't think tonumber is needed. Will remove this.

**Comment by sheharyaar** [comment]
Yes, I am ready

**Review by shreemaan-abhishek — APPROVED** [review]

File: t/plugin/grpc-web.t [hunk 4]
```diff
@@ -68,25 +68,33 @@ passed
...
+Status: { code: 0, details: '', metadata: {} }
 {"name":"hello","path":"/hello"}
+Status: { code: 0, details: '', metadata: {} }
+Status: { code: 0, details: '', metadata: {} }
 {"name":"hello","path":"/hello"}
 
 
 
-=== TEST 3: Proxy server-side streaming request using APISIX gRPC-Web plugin
+=== TEST 3: Proxy server-side streaming request using APISIX with trailers gRPC-Web plugin
```
Comment (shreemaan-abhishek) [inline t/plugin/grpc-web.t:85]: these test cases don't show the trailers, can you add a test case that shows them?

**Review by shreemaan-abhishek — CHANGES_REQUESTED** [review]
https://github.com/apache/apisix/pull/10851/files#r1463630260

File: t/plugin/grpc-web.t [hunk 5]
```diff
@@ -227,3 +235,31 @@ Content-Type: application/grpc-web
...
+
+
+
+=== TEST 12: verify trailers in response
+--- exec
+printf '\u0000\u0000\u0000\u0000\u0007\n\u0005world' | curl 'http://127.0.0.1:1984/grpc/web/a6.RouteService/GetRoute' \
+                      -H 'Accept: */*' \
+                      -H 'Content-Type: application/grpc-web+proto' \
+                      -H 'X-Grpc-Web: 1' \
+                      -H 'X-User-Agent: grpc-web-javascript/0.1' \
```
Comment (shreemaan-abhishek) [inline t/plugin/grpc-web.t:261]: why are the above two headers required?

File: t/plugin/grpc-web.t [hunk 6]
```diff
@@ -227,3 +235,31 @@ Content-Type: application/grpc-web
...
+                      --data-binary '@-' \
+                      --compressed -o -
+--- response_body eval
+qr/grpc-status:0\x0d\x0agrpc-message:/
```
Comment (shreemaan-abhishek) [inline t/plugin/grpc-web.t:265]:
```suggestion
qr/grpc-status:0\x0d\x0agrpc-message:/

```
blank line is needed at the end

**Review by shreemaan-abhishek — CHANGES_REQUESTED** [review]

**Comment by Revolyssup** [comment]
please fix ci

**Comment by sheharyaar** [comment]
I am having issues with this. Locally it works fine for me. I need help with this.

**Review by shreemaan-abhishek — APPROVED** [review]

**Review by monkeyDluffy6017 — APPROVED** [review]

**Pull request merged**